source/lac/sparse_direct.cc

```diff
@@ -928,66 +928,175 @@ SparseDirectMUMPS::initialize_matrix(const Matrix &matrix)
 {
   Assert(matrix.n() == matrix.m(), ExcMessage("Matrix needs to be square."));
 
-  // Hand over matrix to MUMPS as centralized assembled matrix
-  if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+  n    = matrix.n();
+  id.n = n;
+
+  if constexpr (std::is_same_v<SparseMatrix<double>, Matrix>)
     {
-      // Set number of unknowns
-      n = matrix.n();
+      // Serial matrix: hand over matrix to MUMPS as centralized assembled
+      // matrix
+      if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+        {
+          // number of nonzero elements in matrix
+          if constexpr (std::is_same_v<Matrix, SparseMatrix<double>>)
+            nnz = matrix.n_actually_nonzero_elements();
```

Comment:
You don't need the inner `if constexpr` any more here because you already have it on the outside.